src/mscorlib/src/System/Threading/SemaphoreSlim.cs

```diff
@@ -320,6 +323,12 @@ public bool Wait(int millisecondsTimeout, CancellationToken cancellationToken)
 
             cancellationToken.ThrowIfCancellationRequested();
 
+            if (millisecondsTimeout == 0 && m_currentCount == 0)
```

Comment:
Please add a comment about why it's safe to be checking this outside of the lock.